### DISCUSSION TITLE
Mapping using grafonnet

### DISCUSSION THREAD

**Discussion by kelchen123** [grafana/grafonnet#83] [Q&A] [answered]
Hi, I'm currently trying to create mappings for one of my panels using grafonnet, but can't quite figure out how to get started. How could I make a simple value mapping modeled after this grafana json?

"mappings": [
            {
              "options": {
                "123": {
                  "index": 1,
                  "text": "1.2.3"
                }
                "type": "value"
             }
]

**Comment by Duologic** [comment]
The value mappings are documented here: https://grafana.github.io/grafonnet/API/panel/timeSeries/valueMapping.html

**Reply by kelchen123** [reply]
Hi, I tried to use the functions but have been unable to generate any value mapping on grafana. Is there an example format to use? 

For example, I tried formats such as
stateTimeline.valueMapping.RegexMap.options.withPattern("123")
              + stateTimeline.valueMapping.RegexMap.options.withResult("123")
              + stateTimeline.valueMapping.RegexMap.options.result.withText('1.2.3')

**Reply by mrpaulmanley** [reply]
I am struggling to understand this as well. I'm trying to replicate a stat panel that was made through the GUI.
<img width="500" alt="image" src="https://github.com/grafana/grafonnet/assets/385254/73909e7e-875a-42fa-ae25-efafd8df5da3">
It uses the following mappings
<img width="500" alt="image" src="https://github.com/grafana/grafonnet/assets/385254/6736907e-e6c6-4148-a7b6-cf14364c1202">

When I add the following to my panel definition I have 2 problems.
```
      + stat.valueMapping.RangeMap.options.withFrom(0)
      + stat.valueMapping.RangeMap.options.withTo(0)
      + stat.valueMapping.RangeMap.options.result.withText("Down")
      + stat.valueMapping.RangeMap.options.result.withColor("Red")
      + stat.valueMapping.RangeMap.options.withFrom(1)
      + stat.valueMapping.RangeMap.options.withTo(1)
      + stat.valueMapping.RangeMap.options.result.withText("Up")
      + stat.valueMapping.RangeMap.options.result.withColor("Green")
```

1.  Only the last map for range 1-1 is rendered. I'm sure this is just me not fully understanding jsonnet yet.
2. The map is not rendered under the key fieldConfig.defaults.mappings. It instead is rendered as a top level of the panel under options.
```
      {
         "datasource": {
            "type": "datasource",
            "uid": "-- Mixed --"
         },
         "fieldConfig": {
            "defaults": {
               "unit": "s"
            }
         },
         "gridPos": {
            "h": 8,
            "w": 8,
            "x": 0,
            "y": 1
         },
         "id": 2,
         "options": {
            "from": 1,
            "result": {
               "color": "Green",
               "text": "Up"
            },
            "to": 1
         },...
```

Any help would be appreciated.

**Reply by Duologic** [reply]
I intended below comment as reply on this thread.

**Comment by Duologic** [comment, marked as answer]
The value mappings object is a separate object, it can be added to the mappings array of a panel with `standardOptions.withMappings`.

```jsonnet
local valueMapping = star.valueMapping;
stat.standardOptions.withMappings([
  valueMapping.RangeMap.options.withFrom(0)
  + valueMapping.RangeMap.options.withTo(0)
  + valueMapping.RangeMap.options.result.withText('Down')
  + valueMapping.RangeMap.options.result.withColor('Red'),
  valueMapping.RangeMap.options.withFrom(1)
  + valueMapping.RangeMap.options.withTo(1)
  + valueMapping.RangeMap.options.result.withText('Up')
  + valueMapping.RangeMap.options.result.withColor('Green'),
])
```

Documentation for this needs improvement.
